File: core/src/chrislo27/rhre/json/persistent/RemixObject.java
package chrislo27.rhre.json.persistent;

import chrislo27.rhre.track.MusicData;
import com.badlogic.gdx.files.FileHandle;

import java.util.List;

public class RemixObject {

	public transient FileHandle fileHandle;
	public transient MusicData musicData;

	public String musicAssociation;

	public String version;
	public List<EntityObject> entities;

	public float playbackStart;
	public float musicVolume = 1f;
	public float musicStartTime;
	public List<BpmTrackerObject> bpmChanges;

	public MetadataObject metadata = new MetadataObject();

	public static class EntityObject {

		public String id;
		public float beat;
		public int level;

		// optionals
		public boolean isPattern;
		public float width;
		public int semitone;

	}

	public static class BpmTrackerObject {

		public float beat;
		public float tempo;

	}

	public static class MetadataObject {

		public String author;
		public String description;
		public String gamesUsed;

	}

}

File: core/src/chrislo27/rhre/json/persistent/RemixObject.kt
package chrislo27.rhre.json.persistent

import chrislo27.rhre.track.MusicData
import chrislo27.rhre.version.RHRE2Version
import com.badlogic.gdx.files.FileHandle

class RemixObject {

	@Transient var fileHandle: FileHandle? = null
	@Transient var musicData: MusicData? = null

	var musicAssociation: String? = null

	var version: String? = null
	val versionNumber by lazy {
		RHRE2Version.fromString(version ?: throw IllegalStateException("Attempt to compute version number when version is null")).numericalValue
	}
	var entities: MutableList<EntityObject>? = null

	var playbackStart: Float = 0f
	var musicVolume = 1f
	var musicStartTime: Float = 0f
	var bpmChanges: MutableList<BpmTrackerObject>? = null

	var metadata = MetadataObject()

	class EntityObject {

		var id: String? = null
		var beat: Float = 0f
		var level: Int = 0

		// optionals
		var isPattern: Boolean = false
		var width: Float = 0f
		var semitone: Int = 0

		var stopAlways: Boolean = false
		var volume: Float = 1f

	}

	class BpmTrackerObject {

		var beat: Float = 0f
		var tempo: Float = 0f

	}

	class MetadataObject {

		var author: String? = null
		var description: String? = null
		var gamesUsed: String? = null

	}

}
